src/Cardano/Wallet/AddressDerivation.hs

```diff
@@ -0,0 +1,304 @@
+        coinTypeXPrv = -- lvl2 derivation; hardened derivation of coin_type'
+            deriveXPrv DerivationScheme2 pwd purposeXPrv coinTypeIndex
+        acctXPrv = -- lvl3 derivation; hardened derivation of account' index
+            deriveXPrv DerivationScheme2 pwd coinTypeXPrv accIx
+    in
+        Key acctXPrv
+
+-- | Derives address private key from the given account private key, using
+-- derivation scheme 2 (see <https://github.com/input-output-hk/cardano-crypto/ cardano-crypto>
+-- package for more details).
+--
+-- NOTE: The caller is expected to provide the corresponding passphrase. It is
+-- preferred to use 'deriveAddressPublicKey' whenever possible to avoid having
+-- to manipulate passphrases and private keys.
+deriveAddressPrivateKey
+    :: Passphrase "encryption"
+    -> Key 'AccountK XPrv
+    -> ChangeChain
+    -> Index 'Soft 'AddressK
+    -> Key 'AddressK XPrv
+deriveAddressPrivateKey (Passphrase pwd) (Key accXPrv) changeChain (Index addrIx) =
+    let
+        changeCode =
+            fromIntegral $ fromEnum changeChain
+        changeXPrv = -- lvl4 derivation; soft derivation of change chain
+            deriveXPrv DerivationScheme2 pwd accXPrv changeCode
+        addrXPrv = -- lvl5 derivation; soft derivation of address index
+            deriveXPrv DerivationScheme2 pwd changeXPrv addrIx
+    in
+        Key addrXPrv
+
+-- | Derives address public key from the given account public key, using
+-- derivation scheme 2 (see <https://github.com/input-output-hk/cardano-crypto/ cardano-crypto>
+-- package for more details).
+--
+-- This is the preferred way of deriving new sequential address public keys.
+deriveAddressPublicKey
+    :: Key 'AccountK XPub
+    -> ChangeChain
+    -> Index 'Soft 'AddressK
+    -> Key 'AddressK XPub
+deriveAddressPublicKey (Key accXPub) changeChain (Index addrIx) =
+    fromMaybe errWrongIndex $ do
+        let changeCode = fromIntegral $ fromEnum changeChain
+        changeXPub <- -- lvl4 derivation in bip44 is derivation of change chain
+            deriveXPub DerivationScheme2 accXPub changeCode
+        addrXPub <- -- lvl5 derivation in bip44 is derivation of address chain
+            deriveXPub DerivationScheme2 changeXPub addrIx
+        return $ Key addrXPub
+  where
+    errWrongIndex = error $
+        "Cardano.Wallet.AddressDerivation.deriveAddressPublicKey failed: \
+        \was given an hardened (or too big) index for soft path derivation \
+        \( " ++ show addrIx ++ "). This is either a programmer error, or, \
+        \we may have reached the maximum number of addresses for a given \
+        \wallet."
+
+-- | Encode a public key to a (Byron / Legacy) Cardano 'Address'. This is mostly
+-- dubious CBOR serializations with no data attributes.
+keyToAddress
+    :: Key 'AddressK XPub
+    -> Address
+keyToAddress (Key xpub) =
+    Address $ CBOR.toStrictByteString $ encodeAddress xpub encodeAttributes
+  where
+    encodeAttributes = mempty <> CBOR.encodeMapLen 0
```

Comment:
and we do not want to hide encodeAttributes in the implementation of `encodeAddress` and just choose the proper implementation by adding in the `Cardano.Wallet.Binary` : 
```
data AddressScheme = Random | Sequential
```
and just declare what we want at this this level?

Reply:
Nope because the implementation for encoding attributes of random addresses doesn't solely depends on the public key unfortunately. There are extra params that we don't want to bring in the context of the Binary module. The inversion of control is therefore desirable here (cf the prototype).

Reply:
ah, ok, I see now :+1:

Reply:
We have golden tests testing just that actually :) 

--> https://github.com/input-output-hk/cardano-wallet/pull/46/files#diff-9f5308b36f963e411a082134e60e6ebdR76